File: 909. Snakes and Ladders/main.cpp
#include <queue>
#include <vector>

using namespace std;

class Solution {
    static constexpr int ROLL_SIZE = 6;

    void convert(int rowSize, int colSize, int value, int &x, int &y) {
        int r = (value - 1) / colSize;
        int c = (value - 1) % colSize;
        x = rowSize - 1 - r;
        y = (r % 2 == 0) ? c : colSize - 1 - c;
    }

public:
    int snakesAndLadders(vector<vector<int>> &board) {
        queue<pair<int, int>> q;
        q.push({1, 0});
        int rowSize = board.size();
        int colSize = board[0].size();
        int size = rowSize * colSize;
        vector<bool> visited(size + 1, false);
        visited[1] = true;

        while (!q.empty()) {
            int value = q.front().first;
            int moves = q.front().second;
            q.pop();

            for (int i = 1; i <= ROLL_SIZE; i++) {
                int next = value + i;

                if (next > size) {
                    continue;
                }

                int x = 0;
                int y = 0;
                convert(rowSize, colSize, next, x, y);

                if (board[x][y] != -1) {
                    next = board[x][y];
                }

                if (next == size) {
                    return moves + 1;
                }

                if (!visited[next]) {
                    visited[next] = true;
                    q.push({next, moves + 1});
                }
            }
        }

        return -1;
    }
};

File: 909. Snakes and Ladders/main.py
from typing import List, Tuple


class Solution:
    def snakesAndLadders(self, board: List[List[int]]) -> int:
        def convert(rowSize: int, colSize: int, value: int) -> Tuple[int, int]:
            r = (value - 1) // colSize
            c = (value - 1) % colSize
            x = rowSize - 1 - r
            y = c if r % 2 == 0 else colSize - 1 - c

            return (x, y)

        q = [(1, 0)]
        rowSize = len(board)
        colSize = len(board[0])
        size = rowSize * colSize
        visited = [False] * (size + 1)
        visited[1] = True

        while q:
            item = q.pop(0)
            value = item[0]
            moves = item[1]

            for i in range(1, 7):
                next = value + i

                if next > size:
                    continue

                (x, y) = convert(rowSize, colSize, next)

                if board[x][y] != -1:
                    next = board[x][y]

                if next == size:
                    return moves + 1

                if not visited[next]:
                    visited[next] = True
                    q.append((next, moves + 1))

        return -1
